--- ../../common/Makefile	2020-05-22 20:33:39.739723598 +0100
+++ Makefile	2020-05-27 10:47:49.935775223 +0100
@@ -1,12 +1,21 @@
 ###########################################################################################################
 # Compiler, flags and names
 ###########################################################################################################
-CC = gcc
-LD = gcc
-GLOBAL_CC_FLAGS =
-GLOBAL_LD_FLAGS =
+CC = avr-gcc
+LD = avr-gcc
+OC = avr-objcopy
 LIB_D = ./build/deps
 
+CHIP			= atmega328p
+PORT			= /dev/ttyUSB0
+BAUDRATE		= 115200
+
+GLOBAL_CC_FLAGS = -mmcu=$(CHIP)
+GLOBAL_LD_FLAGS = -mmcu=$(CHIP)
+GLOBAL_OC_FLAGS = -R .eeprom
+AVRDUDE_FLAGS	= -p $(CHIP) -c arduino -b $(BAUDRATE) -P $(PORT)
+# May need "-F -V" to stop verification after upload
+
 ###########################################################################################################
 # Makefile includes
 ###########################################################################################################
@@ -17,9 +26,10 @@
 # Main recipe
 ###########################################################################################################
 
-MAIN_BIN 		= a.exe
-MAIN_CC_FLAGS	= $(GLOBAL_CC_FLAGS)
-MAIN_LD_FLAGS	= $(GLOBAL_LD_FLAGS) -L'$(LIB_D)' -Wl,-rpath='$(LIB_D)'
+MAIN_OUT		= a
+MAIN_CC_FLAGS	= $(GLOBAL_CC_FLAGS) -DF_CPU=16000000UL -Os
+MAIN_LD_FLAGS	= $(GLOBAL_LD_FLAGS) 
+MAIN_OC_FLAGS	= $(GLOBAL_OC_FLAGS) -O ihex
 MAIN_SRC_D		= ./src
 MAIN_INC_D		= ./include
 MAIN_OBJ_D		= ./bin/objects
@@ -31,7 +41,9 @@
 	$(CC) $(MAIN_CC_FLAGS) -I'$(MAIN_INC_D)' -c $< -o $@ 
 
 main: $(MAIN_C_OBJ) $(MAIN_INC) # Add any libraries here
-	$(LD) $(MAIN_C_OBJ) $(MAIN_LD_FLAGS) -o $(MAIN_BIN)
+	$(LD) $(MAIN_C_OBJ) $(MAIN_LD_FLAGS) -o $(MAIN_OUT).elf
+	$(OC) $(MAIN_OC_FLAGS) $(MAIN_OUT).elf $(MAIN_OUT).hex
+
 
 ###########################################################################################################
 # General recipes
@@ -43,14 +55,13 @@
 	@echo "Build complete"
 
 # Recipe for running, just builds and executes the binary
-run: build
-	@echo ""	
-	@./$(MAIN_BIN)
+upload: build
+	avrdude $(AVRDUDE_FLAGS) -U flash:w:$(MAIN_OUT).hex
 
 # Recipe for cleaning. Removes all objects and binaries
 clean:
 	rm -rf $(MAIN_OBJ_D)/*.o
-	[ -f "$(MAIN_BIN)" ] && rm $(MAIN_BIN)
+	rm $(MAIN_OUT).*
 	@echo ""
 
 # Recipe for rebuilding. Just an easy way to run "make clean; make build"
